doc/internal/consensus/Consensus-Protocol.md

```diff
@@ -0,0 +1,127 @@
+# The consensus
+
+Before talking about consensus, we need to discuss some logistics based on how the systems can co-exist.
+
+* Communication: Distributed systems need a method to communicate between each other. Remote Procedure Calls is the mechanism using which a standalone server can talk to another. The existing `network` layer of the database will handle all the communication between servers. 
+* Security: Access control mechanisms need to be in place to decide on access to functions in the servers based on their state (leader, follower, candidate)
+* Routing to leader: One of the issues with a varying leader is for the clients to know which IP address to contact for the service. We can solve this problem by advertising any/all IPs of the cluster and simply forward this request to the current leader; OR have a proxy that can forward the request to the current leader wheneve the requests come in. (Section client interaction of post has another approach which works too)
+* The servers will be implemented in the `interal/node` folders which will import the raft API and perform their functions.
+
+Maintaining consensus is one of the major parts of a distributed system. To know to have achieved a stable system, we need the following two parts of implementation.
+
+## The Raft protocol
+
+The raft protocol will be implemented in `internal/raft` and will implement APIs that each node can call.
+
+Raft is an algorithm to handle replicated log, and we maintain the "log" of the SQL stmts applied on a DB and have a completely replicated cluster.
+
+#### General Implementation rules:
+* All RPC calls are done parallely to obtain the best performance.
+* Request retries are done in case of network failures.
+* Raft does not assume network preserves ordering of the packets.
+
+A raft server may be in any of the 3 states; leader, follower or candidate. All requests are serviced through the leader and it then decides how and if the logs must be replicated in the follower machines. The raft protocol has 3 almost independent modules:
```

Comment:
Independent modules means that they live in `internal/raft/{election,replication,safety}`? Or do they all live in `internal/raft`?

Reply:
They will be in a single package `raft` just because I think it'll be easier to use.
By separate modules, I meant they don't interact with each other; though they are in the same package.

Reply:
What do you think?

Reply:
I think that's fine. However, if that bloats the package too much (which I don't think), we should think about moving if possible.